File: coding-interview-array-processing-optimized/src/problems.py
import heapq

# --- Problem 1: Container With Most Water (LeetCode 11) ---

class ContainerWithMostWater:
    """
    Given n non-negative integers a_1, a_2, ..., a_n where each represents a point at
    coordinate (i, a_i). n vertical lines are drawn such that the two endpoints
    of line i is at (i, a_i) and (i, 0). Find two lines, which, together with the x-axis,
    forms a container, such that the container contains the most water.

    Note: You may not slant the container and n is at least 2.
    """

    @staticmethod
    def brute_force(height: list[int]) -> int:
        """
        Approach 1: Brute Force
        Calculates the area for every possible pair of lines and returns the maximum.

        Time Complexity: O(N^2)
            - We use nested loops, where the outer loop iterates from 0 to N-1
              and the inner loop iterates from i+1 to N-1.
              This results in (N-1) + (N-2) + ... + 1 = O(N^2) operations.
        Space Complexity: O(1)
            - We only use a few variables to store the maximum area and loop indices.
        """
        max_area = 0
        n = len(height)
        for i in range(n):
            for j in range(i + 1, n):
                # The width of the container is the distance between the lines.
                width = j - i
                # The height of the container is limited by the shorter line.
                h = min(height[i], height[j])
                # Calculate the area and update max_area if current area is larger.
                max_area = max(max_area, width * h)
        return max_area

    @staticmethod
    def two_pointers_optimal(height: list[int]) -> int:
        """
        Approach 2: Two Pointers (Optimal)
        Uses two pointers, one at the beginning (left) and one at the end (right)
        of the array. In each step, it calculates the area and moves the pointer
        pointing to the shorter line inward. The intuition is that moving the
        shorter line has the potential to increase the height, whereas moving
        the taller line will only decrease the width without guarantee of increasing
        height, as the current shorter line will still be the bottleneck.

        Time Complexity: O(N)
            - The two pointers traverse the array from opposite ends, meeting in the middle.
              Each pointer moves at most N times, resulting in a single pass.
        Space Complexity: O(1)
            - We only use a few variables for pointers, max_area, and current_area.
        """
        max_area = 0
        left = 0
        right = len(height) - 1

        while left < right:
            # Calculate the current width.
            width = right - left
            # Determine the height, limited by the shorter of the two lines.
            h = min(height[left], height[right])
            # Calculate the current area.
            current_area = width * h
            # Update the maximum area found so far.
            max_area = max(max_area, current_area)

            # Move the pointer associated with the shorter line.
            # The goal is to potentially find a taller line to form a larger container.
            if height[left] < height[right]:
                left += 1
            else:
                right -= 1
        return max_area

# --- Problem 2: Product of Array Except Self (LeetCode 238) ---

class ProductOfArrayExceptSelf:
    """
    Given an integer array nums, return an array answer such that answer[i] is equal to
    the product of all the elements of nums except nums[i].
    The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.
    You must write an algorithm that runs in O(n) time and without using the division operation.
    """

    @staticmethod
    def two_arrays_approach(nums: list[int]) -> list[int]:
        """
        Approach 1: Using two auxiliary arrays (Prefix and Suffix Products).
        This approach computes prefix products and suffix products into two separate arrays.
        `prefix_products[i]` stores the product of all elements to the left of `nums[i]`.
        `suffix_products[i]` stores the product of all elements to the right of `nums[i]`.
        Then, `answer[i]` is simply `prefix_products[i] * suffix_products[i]`.

        Time Complexity: O(N)
            - Three passes over the array: one for prefix products, one for suffix products,
              and one to compute the final answer. Each pass takes O(N) time.
        Space Complexity: O(N)
            - Two additional arrays (`prefix_products` and `suffix_products`) of size N are used.
              (If the output array is not counted, this is still O(N) auxiliary space).
        """
        n = len(nums)
        if n == 0:
            return []

        # Initialize prefix and suffix product arrays
        prefix_products = [1] * n
        suffix_products = [1] * n
        answer = [0] * n

        # Calculate prefix products
        # prefix_products[i] = product of nums[0]...nums[i-1]
        for i in range(1, n):
            prefix_products[i] = prefix_products[i-1] * nums[i-1]

        # Calculate suffix products
        # suffix_products[i] = product of nums[i+1]...nums[n-1]
        for i in range(n - 2, -1, -1):
            suffix_products[i] = suffix_products[i+1] * nums[i+1]

        # Calculate the final answer
        # answer[i] = prefix_products[i] * suffix_products[i]
        for i in range(n):
            answer[i] = prefix_products[i] * suffix_products[i]

        return answer

    @staticmethod
    def optimal_o1_space(nums: list[int]) -> list[int]:
        """
        Approach 2: Optimal O(1) space (excluding output array).
        This approach optimizes the space complexity by using the `answer` array itself
        to store prefix products and then calculating suffix products on the fly
        during a second pass.

        1. Initialize `answer` array:
           - First, `answer[i]` will store the product of all elements to the left of `nums[i]`.
           - `answer[0]` is initialized to 1 (nothing to its left).
           - `answer[i] = answer[i-1] * nums[i-1]` for `i > 0`.
        2. Calculate suffix products and combine:
           - Initialize a `right_product` variable to 1. This variable will keep track
             of the product of all elements to the right of the current index `i`.
           - Iterate from `n-1` down to `0`:
             - `answer[i] = answer[i] * right_product` (multiply prefix product by suffix product).
             - `right_product = right_product * nums[i]` (update right product for the next iteration).

        Time Complexity: O(N)
            - Two passes over the array: one forward pass for prefix products
              and one backward pass for suffix products and final computation.
        Space Complexity: O(1)
            - The `answer` array is considered the output and not extra space
              for complexity analysis. Only a few auxiliary variables are used.
        """
        n = len(nums)
        if n == 0:
            return []

        answer = [1] * n

        # Pass 1: Calculate prefix products and store in answer array
        # answer[i] will contain product of nums[0...i-1]
        # For i=0, there are no elements to the left, so answer[0] remains 1.
        for i in range(1, n):
            answer[i] = answer[i-1] * nums[i-1]

        # Pass 2: Calculate suffix products and combine with prefix products
        # 'right_product' will accumulate product of elements to the right
        right_product = 1
        for i in range(n - 1, -1, -1):
            # For the current element answer[i], it already holds the prefix product.
            # We multiply it by the accumulated product of elements to its right.
            answer[i] = answer[i] * right_product
            # Update 'right_product' to include the current element for the next iteration.
            right_product = right_product * nums[i]

        return answer

# --- Problem 3: Rotate Image (Matrix) (LeetCode 48) ---

class RotateImage:
    """
    You are given an n x n 2D matrix representing an image, rotate the image by 90 degrees (clockwise).
    You have to rotate the image in-place, which means you modify the input 2D matrix directly.
    DO NOT allocate another 2D matrix and do the rotation.
    """

    @staticmethod
    def rotate_with_extra_space(matrix: list[list[int]]) -> None:
        """
        Approach 1: Using an auxiliary matrix (Not in-place).
        Creates a new matrix and fills it with rotated elements, then copies back.
        This approach is simple to understand but violates the "in-place" constraint
        of the problem.

        Time Complexity: O(N^2)
            - We iterate through all N*N elements twice (once to copy to new, once to copy back).
        Space Complexity: O(N^2)
            - An entirely new N x N matrix is created to store the rotated image.
        """
        n = len(matrix)
        if n == 0 or n != len(matrix[0]): # Ensure it's a square matrix
            return

        rotated_matrix = [[0] * n for _ in range(n)]

        for r in range(n):
            for c in range(n):
                # A cell (r, c) in the original matrix moves to (c, n - 1 - r) in the rotated matrix.
                rotated_matrix[c][n - 1 - r] = matrix[r][c]

        # Copy the rotated_matrix back to the original matrix
        for r in range(n):
            for c in range(n):
                matrix[r][c] = rotated_matrix[r][c]

    @staticmethod
    def rotate_in_place_transpose_and_reverse(matrix: list[list[int]]) -> None:
        """
        Approach 2: In-place using Transpose and Reverse. (Optimal for clarity/efficiency)
        This method rotates the matrix in two steps:
        1. Transpose the matrix: Swap `matrix[i][j]` with `matrix[j][i]`. This effectively
           flips the matrix over its main diagonal.
        2. Reverse each row: After transposing, reverse the order of elements in each row.

        Example:
        Original:
        1 2 3
        4 5 6
        7 8 9

        Transpose:
        1 4 7
        2 5 8
        3 6 9

        Reverse each row:
        7 4 1
        8 5 2
        9 6 3 (This is the 90-degree clockwise rotation)

        Time Complexity: O(N^2)
            - Transposing takes O(N^2) (iterating over half the matrix).
            - Reversing each row takes O(N^2) (N rows, each reversal is O(N)).
            - Total time is dominated by N^2 operations.
        Space Complexity: O(1)
            - All operations are performed in-place without using additional significant space.
        """
        n = len(matrix)
        if n == 0 or n != len(matrix[0]):
            return

        # Step 1: Transpose the matrix
        # Iterate only for the upper triangle (or lower triangle) to avoid swapping elements twice.
        for r in range(n):
            for c in range(r, n): # Start 'c' from 'r'
                matrix[r][c], matrix[c][r] = matrix[c][r], matrix[r][c]

        # Step 2: Reverse each row
        # After transpose, each row represents a column in the original matrix,
        # but in reverse order for clockwise rotation.
        for r in range(n):
            # Python's list slicing `[::-1]` creates a new list.
            # To reverse in-place, use `list.reverse()` or two pointers.
            matrix[r].reverse() # This modifies the list in-place

    @staticmethod
    def rotate_in_place_layer_by_layer(matrix: list[list[int]]) -> None:
        """
        Approach 3: In-place Layer by Layer Rotation. (Another Optimal O(1) space)
        This method processes the matrix in concentric layers, starting from the outermost.
        For each layer, it rotates the four corresponding elements in a cycle:
        top -> right -> bottom -> left -> top.

        Consider a square matrix of side 'n'. There are `n // 2` layers.
        For each layer `k` (from 0 to `n // 2 - 1`):
        - The elements along the top edge are `(k, c)` where `c` goes from `k` to `n - 1 - k - 1`.
        - The corresponding right edge elements are `(c, n - 1 - k)`.
        - The corresponding bottom edge elements are `(n - 1 - k, n - 1 - c)`.
        - The corresponding left edge elements are `(n - 1 - c, k)`.

        Time Complexity: O(N^2)
            - We iterate through `n // 2` layers.
            - For each layer, we perform `(n - 1 - 2*k)` rotations (where `k` is the layer index).
            - This sums up to approximately N^2 operations.
        Space Complexity: O(1)
            - All operations are performed in-place. Only a temporary variable is used for swapping.
        """
        n = len(matrix)
        if n == 0 or n != len(matrix[0]):
            return

        # Iterate through layers (outermost to innermost)
        # We only need to go up to n // 2 because each layer has 4 sides and a single swap
        # operation handles all 4 sides.
        for layer in range(n // 2):
            first = layer                  # Starting index for the current layer
            last = n - 1 - layer           # Ending index for the current layer

            # Iterate through elements on the top edge of the current layer
            # `i` goes from `first` to `last - 1` (because `last` element is part of the next side segment)
            for i in range(first, last):
                # Store the top element
                top = matrix[first][i]

                # Move left element to top
                matrix[first][i] = matrix[last - (i - first)][first]

                # Move bottom element to left
                matrix[last - (i - first)][first] = matrix[last][last - (i - first)]

                # Move right element to bottom
                matrix[last][last - (i - first)] = matrix[i][last]

                # Move top (stored) element to right
                matrix[i][last] = top

# --- Problem 4: Meeting Rooms II (LeetCode 253) ---

class MeetingRoomsII:
    """
    Given an array of meeting time intervals where intervals[i] = [start_i, end_i],
    return the minimum number of conference rooms required.
    """

    @staticmethod
    def min_rooms_brute_force_conceptual(intervals: list[list[int]]) -> int:
        """
        Approach 1: Brute Force (Conceptual - often too complex to implement in interview)
        This approach would involve iterating through all meetings and for each meeting,
        checking if it can fit into any existing room. If not, a new room is needed.
        Tracking room availability precisely with brute force can become very complex,
        potentially O(N^2) for sorting intervals (if not sorted) and then another O(N^2)
        or O(N*num_rooms) for assigning, making it inefficient.
        A very naive approach might be:
        1. Sort intervals by start time.
        2. For each interval, check against all *active* rooms.
           If it overlaps with all, create a new room.
           This would involve a `List[List[Interval]]` structure where each inner list
           is a room. Checking for overlaps would be slow.
        Given its inefficiency and complexity, this method is usually discussed
        conceptually rather than fully implemented in an interview.
        """
        # This implementation simply provides a base case for the problem,
        # but a true brute-force (simulating rooms and checking all overlaps)
        # is complex and not usually asked to implement for this problem.
        # The simplest (and incorrect for N>1) brute-force might be:
        # if not intervals: return 0
        # if len(intervals) == 1: return 1
        # max_rooms = 0
        # for i in range(len(intervals)):
        #     current_rooms = 1
        #     for j in range(len(intervals)):
        #         if i != j and intervals[i][0] < intervals[j][1] and intervals[j][0] < intervals[i][1]:
        #             current_rooms += 1
        #     max_rooms = max(max_rooms, current_rooms)
        # return max_rooms // 2 if max_rooms > 1 else max_rooms # This is still wrong
        # The correct approach requires tracking simultaneous overlaps, which points to sorting + min-heap.
        print("Brute force for Meeting Rooms II is conceptually difficult and inefficient.")
        print("It's usually not implemented; optimal solutions are preferred.")
        if not intervals:
            return 0
        # A very basic "brute force" that only counts simultaneous overlaps
        # This is more like a conceptual approach and not truly "brute force" in typical interview sense
        # where you try all combinations.
        # This is actually a line sweep, but without a specific data structure to track events.
        # It's less efficient than the optimal heap solution.
        n = len(intervals)
        if n <= 1:
            return n

        # Sort intervals by start time
        intervals.sort(key=lambda x: x[0])

        max_concurrent_meetings = 0
        for i in range(n):
            current_meetings = 1 # The current meeting itself
            for j in range(i + 1, n):
                # Check for overlap: [A_start, A_end], [B_start, B_end]
                # Overlap if A_start < B_end AND B_start < A_end
                # Here, since sorted by start_time, B_start >= A_start
                # So we only need to check B_start < A_end
                if intervals[j][0] < intervals[i][1]:
                    current_meetings += 1
            max_concurrent_meetings = max(max_concurrent_meetings, current_meetings)
        return max_concurrent_meetings


    @staticmethod
    def min_rooms_optimal_heap(intervals: list[list[int]]) -> int:
        """
        Approach 2: Optimal using Sorting and a Min-Heap.
        This is the standard and most efficient approach.

        1. Sort the intervals by their start times.
        2. Initialize a min-heap to store the *end times* of meetings currently in rooms.
           The size of the heap will represent the number of active rooms.
        3. Iterate through the sorted intervals:
           a. If the heap is not empty and the current meeting's start time is
              greater than or equal to the earliest ending meeting (heap[0]),
              it means a room has become free. Remove that meeting's end time from the heap.
           b. Add the current meeting's end time to the heap. This effectively
              assigns the current meeting to a room (either a newly freed one or a new one).
        4. The final size of the heap will be the minimum number of rooms required,
           as it reflects the maximum number of simultaneously active meetings.

        Time Complexity: O(N log N)
            - Sorting the intervals takes O(N log N).
            - Iterating through N intervals: Each heap operation (push, pop) takes O(log K) time,
              where K is the number of elements in the heap (at most N).
            - Total time is dominated by sorting and heap operations: O(N log N + N log N) = O(N log N).
        Space Complexity: O(N)
            - The min-heap can store up to N meeting end times in the worst case
              (e.g., all meetings overlap).
        """
        if not intervals:
            return 0

        # 1. Sort intervals by their start times
        intervals.sort(key=lambda x: x[0])

        # 2. Initialize a min-heap to store the end times of meetings
        # The heap will effectively store the ending times of meetings that are
        # currently occupying rooms.
        # The top (smallest) element of the heap will be the earliest ending meeting.
        rooms = [] # min-heap

        # 3. Iterate through the sorted intervals
        for start, end in intervals:
            # Check if any room is free (i.e., if the earliest ending meeting has finished
            # before the current meeting starts).
            if rooms and start >= rooms[0]:
                # If a room is free, reuse it. Remove the earliest ending meeting.
                heapq.heappop(rooms)
            # Assign the current meeting to a room (either new or reused).
            # Add its end time to the heap.
            heapq.heappush(rooms, end)

        # 4. The number of elements in the heap is the minimum rooms required.
        # This is because the heap always contains the end times of currently
        # active meetings, and its size reflects the maximum concurrency.
        return len(rooms)

    @staticmethod
    def min_rooms_line_sweep(intervals: list[list[int]]) -> int:
        """
        Approach 3: Line Sweep (Alternative Optimal)
        This approach uses a "line sweep" or "event point" technique.
        1. Create two lists: one for all start times and one for all end times.
        2. Sort both lists independently.
        3. Iterate through the sorted start times and sorted end times simultaneously
           using two pointers.
           - If a meeting starts (`start_ptr` advances), increment a `rooms_needed` counter.
           - If a meeting ends (`end_ptr` advances), decrement `rooms_needed`.
           - The maximum value `rooms_needed` reaches is the answer.
        Crucially, if a meeting ends *at the same time* another starts, we prioritize
        releasing the room first (decrement then increment) so that the same room can be reused.
        This means if `start_times[s_ptr] == end_times[e_ptr]`, advance `e_ptr` first.

        Time Complexity: O(N log N)
            - Creating and sorting `start_times` and `end_times` lists takes O(N log N).
            - The two-pointer sweep takes O(N).
            - Total: O(N log N).
        Space Complexity: O(N)
            - Two additional lists of size N are created for start and end times.
        """
        if not intervals:
            return 0

        n = len(intervals)
        start_times = sorted([interval[0] for interval in intervals])
        end_times = sorted([interval[1] for interval in intervals])

        rooms_needed = 0
        max_rooms = 0
        s_ptr = 0 # Pointer for start times
        e_ptr = 0 # Pointer for end times

        while s_ptr < n:
            # If the current meeting starts before or at the same time as
            # the earliest ending meeting, a new room is needed (or a room
            # is just freed and immediately taken).
            # Prioritize ending a meeting if start time == end time
            if start_times[s_ptr] < end_times[e_ptr]:
                rooms_needed += 1
                s_ptr += 1
            else: # A meeting has ended, so a room is freed
                rooms_needed -= 1
                e_ptr += 1

            # Update the maximum rooms needed so far
            max_rooms = max(max_rooms, rooms_needed)

        return max_rooms
---

File: leetcode-linked-lists-solutions-complete/src/main/java/com/example/linkedlist/Problems.java
```java
package com.example.linkedlist;

import java.util.Comparator;
import java.util.PriorityQueue;

/**
 * This class contains optimal solutions for various classic Linked List problems.
 * Each method includes detailed comments on logic, time, and space complexity.
 */
public class Problems {

    /**
     * Problem 1: Reorder List
     * Given a singly linked list L: L0 -> L1 -> ... -> Ln-1 -> Ln,
     * reorder it to: L0 -> Ln -> L1 -> Ln-1 -> L2 -> Ln-2 -> ...
     * You may not modify the values in the list's nodes. Only nodes themselves may be changed.
     *
     * Example:
     * Input: head = [1,2,3,4]
     * Output: [1,4,2,3]
     *
     * Example:
     * Input: head = [1,2,3,4,5]
     * Output: [1,5,2,4,3]
     *
     * Optimal Approach:
     * 1. Find the middle of the linked list. (e.g., using fast and slow pointers).
     * 2. Split the linked list into two halves. The second half starts from (middle.next).
     *    The first half ends at middle, so set middle.next = null.
     * 3. Reverse the second half of the linked list.
     * 4. Merge the two halves alternating nodes.
     *
     * Time Complexity: O(N)
     *   - Finding middle: O(N)
     *   - Reversing second half: O(N/2) = O(N)
     *   - Merging: O(N/2) = O(N)
     *   Overall: O(N)
     *
     * Space Complexity: O(1)
     *   - Only a few pointers are used, no extra data structures proportional to N.
     */
    public void reorderList(LinkedListNode head) {
        if (head == null || head.next == null) {
            return; // List is empty or has only one node, no reordering needed.
        }

        // Step 1: Find the middle of the linked list using fast and slow pointers.
        // When fast reaches the end, slow will be at the middle.
        LinkedListNode slow = head;
        LinkedListNode fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        // At this point, 'slow' is the middle node.
        // If the list has an odd number of nodes, slow is the actual middle.
        // If the list has an even number of nodes, slow is the first of the two middle nodes.

        // Step 2: Split the linked list into two halves.
        // first half: head -> ... -> (slow node before split)
        // second half: (slow.next before split) -> ... -> null
        LinkedListNode firstHalfHead = head;
        LinkedListNode secondHalfHead = slow.next; // The head of the second half
        slow.next = null; // Break the list into two distinct halves

        // Step 3: Reverse the second half of the linked list.
        secondHalfHead = reverseList(secondHalfHead);

        // Step 4: Merge the two halves alternating nodes.
        // Example: L1: 1 -> 2 -> 3
        //          L2: 6 -> 5 -> 4 (reversed from 4 -> 5 -> 6)
        // Result: 1 -> 6 -> 2 -> 5 -> 3 -> 4
        LinkedListNode p1 = firstHalfHead;
        LinkedListNode p2 = secondHalfHead;
        while (p1 != null && p2 != null) {
            LinkedListNode p1Next = p1.next; // Store next node of first half
            LinkedListNode p2Next = p2.next; // Store next node of second half

            p1.next = p2;       // Link first half node to second half node
            p2.next = p1Next;   // Link second half node to next first half node

            p1 = p1Next;        // Move p1 to its original next
            p2 = p2Next;        // Move p2 to its original next
        }
        // If one list is longer (only possible for the first half if odd length original list),
        // the remaining nodes are already correctly linked at the end of the merged list.
        // For example, if original list [1,2,3,4,5], first half [1,2,3], second half reversed [5,4].
        // Merging: 1->5->2->4->3. '3' remains linked to null naturally.
    }

    /**
     * Helper function to reverse a linked list.
     * This is a standard iterative approach.
     *
     * @param head The head of the list to be reversed.
     * @return The new head of the reversed list.
     */
    private LinkedListNode reverseList(LinkedListNode head) {
        LinkedListNode prev = null;
        LinkedListNode current = head;
        while (current != null) {
            LinkedListNode nextTemp = current.next; // Store next node
            current.next = prev;                   // Reverse current node's pointer
            prev = current;                        // Move prev to current node
            current = nextTemp;                    // Move current to next node
        }
        return prev; // 'prev' will be the new head (original tail)
    }

    /**
     * Problem 2: Add Two Numbers
     * You are given two non-empty linked lists representing two non-negative integers.
     * The digits are stored in reverse order, and each of their nodes contains a single digit.
     * Add the two numbers and return the sum as a linked list.
     * You may assume the two numbers do not contain any leading zero, except the number 0 itself.
     *
     * Example:
     * Input: l1 = [2,4,3], l2 = [5,6,4]  (represents 342 + 465)
     * Output: [7,0,8]                     (represents 807)
     *
     * Example:
     * Input: l1 = [9,9,9,9,9,9,9], l2 = [9,9,9,9]
     * Output: [8,9,9,9,0,0,0,1]
     *
     * Optimal Approach:
     * Iterate through both lists simultaneously, summing digits and handling carry.
     * Create a new linked list for the result.
     *
     * Time Complexity: O(max(M, N)), where M and N are the lengths of l1 and l2 respectively.
     *   - We traverse each list at most once.
     *
     * Space Complexity: O(max(M, N))
     *   - A new linked list is created to store the sum. In the worst case (e.g., 99 + 1 = 100),
     *     the sum list can be one node longer than the longest input list.
     */
    public LinkedListNode addTwoNumbers(LinkedListNode l1, LinkedListNode l2) {
        // Dummy head for the result list, simplifies adding the first node.
        LinkedListNode dummyHead = new LinkedListNode(0);
        LinkedListNode current = dummyHead; // Pointer to build the result list
        int carry = 0; // Carry-over from previous sum

        // Iterate as long as there are digits in either list or a carry exists.
        while (l1 != null || l2 != null || carry != 0) {
            int val1 = (l1 != null) ? l1.val : 0; // Get value from l1, or 0 if l1 is null
            int val2 = (l2 != null) ? l2.val : 0; // Get value from l2, or 0 if l2 is null

            int sum = val1 + val2 + carry; // Calculate sum of digits and carry
            carry = sum / 10;              // Update carry (e.g., 17 / 10 = 1)
            int digit = sum % 10;          // Get the digit for the current node (e.g., 17 % 10 = 7)

            current.next = new LinkedListNode(digit); // Create new node with the digit
            current = current.next;                   // Move current pointer forward

            if (l1 != null) {
                l1 = l1.next; // Move l1 forward if not null
            }
            if (l2 != null) {
                l2 = l2.next; // Move l2 forward if not null
            }
        }

        return dummyHead.next; // The actual head of the sum list is after the dummy node.
    }

    /**
     * Problem 3: Merge k Sorted Lists
     * You are given an array of k linked-lists lists, each linked list is sorted in ascending order.
     * Merge all the linked-lists into one sorted linked list and return it.
     *
     * Example:
     * Input: lists = [[1,4,5],[1,3,4],[2,6]]
     * Output: [1,1,2,3,4,4,5,6]
     * Explanation: The linked lists are:
     * [1->4->5],
     * [1->3->4],
     * [2->6]
     * merging them into one sorted list: 1->1->2->3->4->4->5->6
     *
     * Optimal Approach: Using a Min-Priority Queue (Min-Heap).
     * 1. Create a Min-Priority Queue that stores LinkedListNodes, ordered by their 'val'.
     * 2. Add the head of each non-empty list into the priority queue.
     * 3. Repeatedly extract the minimum node from the PQ. This node is the next node in our merged list.
     * 4. If the extracted node has a 'next' node, add that 'next' node to the PQ.
     * 5. Continue until the PQ is empty.
     *
     * Time Complexity: O(N log k)
     *   - N is the total number of nodes across all k lists.
     *   - k is the number of linked lists.
     *   - Each node is added to the priority queue once and extracted once.
     *   - Adding/extracting from a priority queue of size k takes O(log k) time.
     *   - Total operations: N * O(log k).
     *
     * Space Complexity: O(k)
     *   - The priority queue stores at most k nodes (one from each list).
     */
    public LinkedListNode mergeKLists(LinkedListNode[] lists) {
        if (lists == null || lists.length == 0) {
            return null;
        }

        // Create a min-priority queue to store the head nodes of each list.
        // It will order nodes based on their 'val' in ascending order.
        PriorityQueue<LinkedListNode> minHeap = new PriorityQueue<>(Comparator.comparingInt(node -> node.val));

        // Add the head of each non-null list to the priority queue.
        for (LinkedListNode listHead : lists) {
            if (listHead != null) {
                minHeap.add(listHead);
            }
        }

        // Dummy head for the merged list, simplifies construction.
        LinkedListNode dummyHead = new LinkedListNode(0);
        LinkedListNode tail = dummyHead; // Pointer to the last node of the merged list

        // While the priority queue is not empty, extract the smallest node.
        while (!minHeap.isEmpty()) {
            LinkedListNode smallestNode = minHeap.poll(); // Get and remove the smallest node

            tail.next = smallestNode; // Append the smallest node to the merged list
            tail = tail.next;         // Move the tail pointer forward

            // If the extracted node has a next node, add it to the priority queue.
            // This ensures we always have the next available candidate from that list.
            if (smallestNode.next != null) {
                minHeap.add(smallestNode.next);
            }
        }

        return dummyHead.next; // The actual head of the merged list.
    }

    /**
     * Problem 4: Reverse Nodes in k-Group
     * Given the head of a linked list, reverse the nodes of the list k at a time,
     * and return the modified list.
     * k is a positive integer and is less than or equal to the length of the linked list.
     * If the number of nodes is not a multiple of k, then the remaining nodes,
     * in the end, should remain as they are.
     * You may not alter the values in the list's nodes, only nodes themselves may be changed.
     *
     * Example:
     * Input: head = [1,2,3,4,5], k = 2
     * Output: [2,1,4,3,5]
     *
     * Example:
     * Input: head = [1,2,3,4,5], k = 3
     * Output: [3,2,1,4,5]
     *
     * Optimal Approach: Iterative reversal of k-length segments.
     * This involves careful pointer manipulation to connect the reversed segments.
     *
     * 1. Use a dummy node to simplify handling the new head.
     * 2. Traverse the list, identifying k-node segments.
     * 3. For each k-node segment:
     *    a. Check if there are k nodes remaining. If not, break (or just append the rest).
     *    b. Reverse the k nodes.
     *    c. Connect the reversed segment back to the previous segment.
     *    d. The original head of the k-segment becomes the tail of the reversed segment.
     *       This tail then points to the next segment's head.
     *
     * Time Complexity: O(N)
     *   - Each node is visited and its `next` pointer is modified a constant number of times.
     *
     * Space Complexity: O(1)
     *   - Only a few pointers are used for iteration and reversal. No extra data structures
     *     proportional to N or k are created.
     */
    public LinkedListNode reverseKGroup(LinkedListNode head, int k) {
        if (head == null || k == 1) {
            return head; // No reversal needed for null list or k=1
        }

        // Dummy node to simplify handling the new head of the list.
        LinkedListNode dummy = new LinkedListNode(0);
        dummy.next = head;

        LinkedListNode prevGroupTail = dummy; // The node before the current k-group
        LinkedListNode current = head;        // The current node being processed

        while (current != null) {
            // Step 1: Check if there are at least k nodes remaining in the current segment.
            LinkedListNode groupHead = current; // Head of the current k-group
            LinkedListNode groupEnd = current;  // Will become the end of the current k-group

            int count = 0;
            while (groupEnd != null && count < k) {
                groupEnd = groupEnd.next;
                count++;
            }

            // If we don't have k nodes, we stop and don't reverse the remaining nodes.
            if (count < k) {
                break;
            }

            // Step 2: Reverse the k-group.
            // 'groupHead' is the start of the segment to reverse.
            // 'groupEnd' is the node *after* the segment to reverse (or null if it's the end of list).
            LinkedListNode reversedHead = reverseSublist(groupHead, groupEnd); // Reverses from groupHead up to (but not including) groupEnd

            // Step 3: Connect the reversed group.
            // The `prevGroupTail` (which initially is dummy) should point to the new `reversedHead`.
            // The original `groupHead` (which is now the tail of the reversed segment) should point to `groupEnd`.
            prevGroupTail.next = reversedHead;   // Connect the previous group to the new head of the reversed group
            groupHead.next = groupEnd;           // Connect the tail of the reversed group to the next group's head

            // Step 4: Update pointers for the next iteration.
            prevGroupTail = groupHead; // The tail of the current reversed group becomes the prevGroupTail for the next.
            current = groupEnd;        // The current node for the next iteration starts from where this group ended.
        }

        return dummy.next; // The new head of the entire modified list.
    }

    /**
     * Helper function to reverse a sublist of a linked list.
     * Reverses the list from 'start' up to (but not including) 'end'.
     *
     * @param start The first node of the sublist to reverse.
     * @param end   The node AFTER the last node of the sublist to reverse.
     *              The reversal stops just before 'end'.
     * @return The new head of the reversed sublist (which was 'start's k-th node).
     */
    private LinkedListNode reverseSublist(LinkedListNode start, LinkedListNode end) {
        LinkedListNode prev = null;
        LinkedListNode current = start;
        // Keep reversing until 'current' reaches 'end'.
        while (current != end) {
            LinkedListNode nextTemp = current.next;
            current.next = prev;
            prev = current;
            current = nextTemp;
        }
        return prev; // 'prev' is the new head of the reversed sublist.
    }
}
```